### ISSUE TITLE
.NET 7 exception on Android with EnableLLVM=true

### ISSUE THREAD

**Issue by jonathanpeppers** [dotnet/runtime#73003]
### Description

Run with LLVM on an arm64 device, and you can run into:
```
07-22 09:59:55.062  4374  4374 F mono-rt : [ERROR] FATAL UNHANDLED EXCEPTION: System.PlatformNotSupportedException: Arg_PlatformNotSupported
07-22 09:59:55.062  4374  4374 F mono-rt :    at System.Text.ASCIIUtility.GetIndexOfFirstNonAsciiByte_Intrinsified(Byte* , UIntPtr )
07-22 09:59:55.062  4374  4374 F mono-rt :    at System.Text.Unicode.Utf8Utility.GetPointerToFirstInvalidByte(Byte* , Int32 , Int32& , Int32& )
07-22 09:59:55.062  4374  4374 F mono-rt :    at System.Text.UTF8Encoding.GetCharCount(Byte* , Int32 )
07-22 09:59:55.062  4374  4374 F mono-rt :    at System.String.CreateStringFromEncoding(Byte* , Int32 , Encoding )
07-22 09:59:55.062  4374  4374 F mono-rt :    at System.Text.Encoding.GetString(Byte* , Int32 )
07-22 09:59:55.062  4374  4374 F mono-rt :    at System.String.CreateStringForSByteConstructor(Byte* , Int32 )
07-22 09:59:55.062  4374  4374 F mono-rt :    at System.String.Ctor(SByte* value)
07-22 09:59:55.062  4374  4374 F mono-rt :    at Java.Interop.TypeManager.GetClassName(IntPtr )
07-22 09:59:55.062  4374  4374 F mono-rt :    at Android.Runtime.JNIEnv.RegisterJniNatives(IntPtr , Int32 , IntPtr , IntPtr , Int32 )
```

Somewhere here:

https://github.com/dotnet/runtime/blob/b31a48ae77cf20729387383d15e207f838b0a076/src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs#L88

So I think `AdvSimd.Arm64.IsSupported` is `true`, but because `GetIndexOfFirstNonAsciiByte_Intrinsified()` is not in the AOT profile (it's running under JIT), we get the above exception. It's possible that the above code should also work under JIT.

/cc @fanyang-mono 

### Reproduction Steps

1. `dotnet new android`
2. `dotnet build -t:Run -c Release -p:EnableLLVM=true` to run on an arm64 device (in my case a Pixel 5)

### Expected behavior

You can use Profiled AOT + LLVM together on Android.

### Actual behavior

We get an exception when trying to use Profiled AOT + LLVM together on Android.

### Regression?

I don't hit this same exception in .NET 6, but maybe it's just luck?

### Known Workarounds

You can turn off profiled AOT, such as `-p:AndroidEnableProfiledAot=false` or don't use LLVM.

### Configuration

.NET 7.0.100-rc.1.22374.1

**Comment by jonathanpeppers** [comment]
We also tried passing `-O=-simd` to the AOT compiler, but ran into a different issue.

But that might be solved by: https://github.com/dotnet/runtime/pull/72982

**Comment by steveisok** [comment]
@vargaz Will https://github.com/dotnet/runtime/pull/72982 solve this?

**Comment by vargaz** [comment]
Hopefully.

**Comment by steveisok** [comment]
@jonathanpeppers Any chance you can help us validate?

**Comment by jonathanpeppers** [comment]
I triggered Maestro just now, but I don't see the change there yet: https://github.com/xamarin/xamarin-android/pull/7217

Maybe we can try it tomorrow.

**Comment by fanyang-mono** [comment]
This could be a dup of https://github.com/dotnet/runtime/issues/60792

**Comment by lambdageek** [comment]
this is a nasty problem.

so the issue is.

1. Android uses profiled AOT pretty much by default now - for fast startup
2. There are functions that don't get AOTed (they're not early enough or they weren't used in the profile that Android uses)
3. when you turn on LLVM, the method that makes the decision "should we support intrinsics" gets inlined into something that is part of the AOT profile, so `AdvSimd.IsSupported` is hardcoded to "true"
https://github.com/dotnet/runtime/blob/aae6e9b0759ddce43cffa9d059f0cea42bb3ba54/src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs#L81
4. But at runtime the actual method that does the work is JITed, so when it checks `AdvSimd.IsSupported` it gets false 
https://github.com/dotnet/runtime/blob/aae6e9b0759ddce43cffa9d059f0cea42bb3ba54/src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs#L294

I believe the fix is that when we check for "jit_supported" instead of just checking "are we using LLVM", we need to check "and we don't have the JIT as a fallback".  in other words, we have to be in FullAOT or HybridAOT mode.

https://github.com/dotnet/runtime/blob/8027c80a36ae73f2c729c6950c695a418304ecf3/src/mono/mono/mini/simd-intrinsics.c#L721-L723

In other words: profiled AOT cannot benefit from the arm64 intrinsics work we've done for LLVM.

**Comment by akoeplinger** [comment]
not a fix for the underlying issue but could we somehow force the intrinsics APIs to be included in the profile so that we still get the LLVM intrinsics handling by default?

**Comment by lambdageek** [comment]
> not a fix for the underlying issue but could we somehow force the intrinsics APIs to be included in the profile so that we still get the LLVM intrinsics handling by default?

I'm not sure I understand the suggestion.

The issue is:  the code that decides if we should use an optimized version of an algorithm sees different `IsSupported` values than the code that implements the specialized version of the algorithm.

The code that decides is compiled with LLVM AOT and therefore `IsSupported` is true.  The code that implements the algorithm is JITed and locally sees `IsSupported` as false.

**Comment by lambdageek** [comment]
Basically the "fallback" execution engines need to return `IsSupported == true` if any accelerted execution engine returns `IsSupported == true`.  Otherwise code is unsafe.

**Comment by fanyang-mono** [comment]
The reason why Android is seeing this is because Android uses this setup: LLVM AOT with non-LLVM JIT fallback. Currently, `AdvSimd.IsSupported` only returns `true`, when LLVM is enabled, no matter it is AOT or JIT.

**Comment by akoeplinger** [comment]
> The issue is: the code that decides if we should use an optimized version of an algorithm sees different IsSupported values than the code that implements the specialized version of the algorithm.

My suggestion was to make sure we _always_ LLVM AOT the implementation, regardless of whether it is used by the profile, so that no fallback to the JIT occurs.

**Comment by lambdageek** [comment]
> > The issue is: the code that decides if we should use an optimized version of an algorithm sees different IsSupported values than the code that implements the specialized version of the algorithm.
> 
> My suggestion was to make sure we _always_ LLVM AOT the implementation, regardless of whether it is used by the profile, so that no fallback to the JIT occurs.

That's hard for a number of reasons:
1. We don't know what "the implementation" is. In this case it's called after the IsSupported check, but in general the decision about which algorithm to use could be far removed from the place where the intrinsics get used.
2. The implementation method could use something that we don't know how to AOT like exception filters.

If we had some attribute `[ForceAOT(AdvSimd.IsSupported)]` then maybe we could address (1), and then (2) could be an error and we could educate users to only use the supported constructs or refactor their code
